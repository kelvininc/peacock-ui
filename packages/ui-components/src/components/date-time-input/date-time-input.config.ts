export const DATE_TIME_INPUTMASK_CONFIG: Inputmask.Options = {
	alias: 'datetime',
	inputFormat: 'dd-mm-yyyy HH:MM:ss',
	displayFormat: 'dd-mm-yyyy 00:00:00',
	placeholder: 'dd-mm-yyyy 00:00:00',
	prefillYear: false,
	showMaskOnHover: false,
	clearMaskOnLostFocus: false
};
