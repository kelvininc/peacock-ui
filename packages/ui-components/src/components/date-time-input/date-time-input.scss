@import '../../assets/styles/globals';

kv-form-label {
	--label-color: #{kv-color('neutral-4')};
}

input {
	@include ellipsis;
	@include kv-font-span-regular;

	color-scheme: dark;
	appearance: textfield;
	outline: 0;
	background: kv-color('neutral-6');
	border: transparent;
	padding: 0 $spacing-2x;
	width: 100%;
}

.input-container-wrapper {
	display: flex;
	align-items: center;
	background-color: kv-color('neutral-6');
	min-width: 196px;

	&--size-large {
		height: 36px;
	}

	&--size-small {
		height: 28px;
	}

	&--style-merged-left {
		border-top-left-radius: 4px;
		border-bottom-left-radius: 4px;
	}

	&--style-merged-right {
		border-top-right-radius: 4px;
		border-bottom-right-radius: 4px;
	}

	&--style-separated {
		border-radius: 4px;
	}

	.input-separator {
		width: 1px;
		height: 20px;
		background-color: kv-color('neutral-5');
	}
}

.input-container {
	position: relative;
	display: flex;
	align-items: center;
	justify-content: space-between;
	height: 100%;
	width: 100%;
	border-radius: 4px;
	border: solid 1px kv-color('neutral-6');
	color: kv-color('neutral-2');
	overflow: hidden;
	box-sizing: border-box;

	kv-icon {
		--icon-color: #{kv-color('neutral-4')};
	}

	&.focus,
	&.forced-focus {
		color: kv-color('neutral-2');
		border-color: kv-color('neutral-0');


	kv-icon {
		--icon-color: #{kv-color('neutral-2')};
	}

		&::placeholder {
			color: kv-color('neutral-4');
		}
	}

	&.disabled {
		background-color: kv-color('neutral-7');
		color: kv-color('neutral-5');

		kv-icon {
			--icon-color: #{kv-color('neutral-5')};
		}
	}

	&.invalid {
		border-color: kv-color('error');
	}

	.left-container {
		display: flex;
		align-items: center;
		width: 100%;

		.left-icon {
			padding-left: $spacing-2x;

			kv-icon {
				--icon-height: 20px;
				--icon-width: 20px;
			}
		}
	}

	.right-icon {
		padding-right: $spacing-2x;
		cursor: pointer;

		kv-icon {
			--icon-height: 20px;
			--icon-width: 20px;
		}
	}
}
