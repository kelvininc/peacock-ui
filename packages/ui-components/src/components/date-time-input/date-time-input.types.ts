import { EventEmitter } from '@stencil/core';
import { EComponentSize, EValidationState } from '../../types';

export interface IDateTimeInputEvents {
	/** Emitted when a keyboard input occurred */
	textChange: EventEmitter<string>;
	/** Emitted when date time lost focus */
	dateTimeBlur: EventEmitter<string>;
	/** Emitted when the input is foccused */
	inputFocus: EventEmitter<FocusEvent>;
}

export interface IDateTimeInput {
	/** (optional) Date time input label */
	label?: string;
	/** (optional) Date time input name */
	inputName?: string;
	/** (optional) Date time place holder */
	placeholder?: string;
	/** (optional) Sets this tab item to a different styling configuration */
	size: EComponentSize;
	/** (optional) Date time value */
	value?: string | number | null;
	/** (optional) Date time focus state */
	forcedFocus?: boolean;
	/** (optional) Similar to forcedFocus but does not emmit events */
	highlighted?: boolean;
	/** (optional) Use a input mask when the Date time type is a Datetime (default true) */
	useInputMask?: boolean;
	/** (optional) Date time input disabled */
	disabled?: boolean;
	/** (optional) Date time input required */
	required?: boolean;
	/** (optional) Date time input state */
	state?: EValidationState;
	/** (optional) Date time input help text */
	helpText?: string | string[];
}
