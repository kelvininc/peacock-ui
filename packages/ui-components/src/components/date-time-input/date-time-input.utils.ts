import { isEmpty } from 'lodash';

/**
 * This inputmask cannot handle date limits if one of the limit is not defined.
 * Therefore, if we need to use this feature, we should always set the two limits.
 */
export const getDateLimitsInputmaskConfig = (min?: string, max?: string) => {
	if (!isEmpty(min) && !isEmpty(max)) {
		return {
			min,
			max
		};
	}
};
