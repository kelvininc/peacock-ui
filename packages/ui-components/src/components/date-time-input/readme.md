# kv-date-time-input



<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-date-time-input></kv-date-time-input>

<!-- Labeled -->
<kv-date-time-input label="Text Field"></kv-date-time-input>
```


### React

```tsx
import React from 'react';

import { KvDateTimeInput } from '@kelvininc/react-ui-components';

export const DateTimeInputExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvDateTimeInput />

		{/*-- Labeled --*/}
		<KvDateTimeInput label="Text Field" />
	</>
);
```



## Properties

| Property       | Attribute        | Description                                                                      | Type                                                                          | Default                 |
| -------------- | ---------------- | -------------------------------------------------------------------------------- | ----------------------------------------------------------------------------- | ----------------------- |
| `disabled`     | `disabled`       | (optional) Date time input disabled                                              | `boolean`                                                                     | `false`                 |
| `forcedFocus`  | `forced-focus`   | (optional) Date time focus state                                                 | `boolean`                                                                     | `false`                 |
| `helpText`     | `help-text`      | (optional) Date time input help text                                             | `string \| string[]`                                                          | `[]`                    |
| `highlighted`  | `highlighted`    | (optional) Similar to forcedFocus but does not emmit events                      | `boolean`                                                                     | `false`                 |
| `inputName`    | `input-name`     | (optional) Date time input name                                                  | `string`                                                                      | `undefined`             |
| `label`        | `label`          | (optional) Date time input label                                                 | `string`                                                                      | `undefined`             |
| `placeholder`  | `placeholder`    | (optional) Date time place holder                                                | `string`                                                                      | `''`                    |
| `required`     | `required`       | (optional) Date time input required                                              | `boolean`                                                                     | `false`                 |
| `size`         | `size`           | (optional) Sets this tab item to a different styling configuration               | `EComponentSize.Large \| EComponentSize.Small`                                | `EComponentSize.Large`  |
| `state`        | `state`          | (optional) Date time input state                                                 | `EValidationState.Invalid \| EValidationState.None \| EValidationState.Valid` | `EValidationState.None` |
| `useInputMask` | `use-input-mask` | (optional) Use a input mask when the Date time type is a Datetime (default true) | `boolean`                                                                     | `false`                 |
| `value`        | `value`          | (optional) Date time value                                                       | `string`                                                                      | `''`                    |


## Events

| Event          | Description                            | Type                      |
| -------------- | -------------------------------------- | ------------------------- |
| `dateTimeBlur` | Emitted when date time lost focus      | `CustomEvent<string>`     |
| `inputFocus`   | Emitted when the input is foccused     | `CustomEvent<FocusEvent>` |
| `textChange`   | Emitted when a keyboard input occurred | `CustomEvent<string>`     |


## Dependencies

### Used by

 - [kv-absolute-time-picker](../absolute-time-picker)

### Depends on

- [kv-form-label](../form-label)
- [kv-form-help-text](../form-help-text)

### Graph
```mermaid
graph TD;
  kv-date-time-input --> kv-form-label
  kv-date-time-input --> kv-form-help-text
  kv-form-help-text --> kv-icon
  kv-absolute-time-picker --> kv-date-time-input
  style kv-date-time-input fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


