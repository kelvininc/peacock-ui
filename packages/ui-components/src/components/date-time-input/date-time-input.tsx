import { Component, Element, Event, EventEmitter, Fragment, Host, Prop, State, Watch, h } from '@stencil/core';
import { EInputFieldType, EValidationState } from '../text-field/text-field.types';
import { EComponentSize } from '../../types';
import { isNil } from 'lodash';
import { DATE_TIME_INPUTMASK_CONFIG } from './date-time-input.config';
import { IDateTimeInput, IDateTimeInputEvents } from './date-time-input.types';
import Inputmask from 'inputmask';

@Component({
	tag: 'kv-date-time-input',
	styleUrl: 'date-time-input.scss',
	shadow: false
})
export class KvDateTimeInput implements IDateTimeInput, IDateTimeInputEvents {
	@Element() el!: HTMLKvDateTimeInputElement;

	/** @inheritdoc */
	@Prop({ reflect: true }) label?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) inputName?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) placeholder?: string = '';
	/** @inheritdoc */
	@Prop({ reflect: true }) value?: string | null = '';
	/** @inheritdoc */
	@Prop({ reflect: true }) useInputMask?: boolean = false;
	/** @inheritdoc */
	@Prop() size: EComponentSize = EComponentSize.Large;
	/** @inheritdoc */
	@Prop({ reflect: true }) forcedFocus: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) highlighted: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) required: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) state: EValidationState = EValidationState.None;
	/** @inheritdoc */
	@Prop({ reflect: true }) helpText: string | string[] = [];

	@State() focused = false;

	private nativeInput?: HTMLInputElement;

	/** @inheritdoc */
	@Event() textChange: EventEmitter<string>;
	/** @inheritdoc */
	@Event() dateTimeBlur: EventEmitter<string>;
	/** @inheritdoc */
	@Event() inputFocus: EventEmitter<FocusEvent>;

	@Watch('forcedFocus')
	forcedFocusChangeHandler(newValue: boolean) {
		this.focused = newValue;

		if (!this.focused) {
			this.el.blur();
		}
	}

	@Watch('useInputMask')
	handleUseInputMask(useInputMaskValue: boolean) {
		if (useInputMaskValue) {
			this.createInputMaskInstance();
		} else if (this.nativeInput) {
			Inputmask.remove(this.nativeInput);
		}
	}

	componentWillLoad() {
		this.focused = this.forcedFocus;
	}

	componentDidLoad() {
		this.handleUseInputMask(this.useInputMask);
	}

	private createInputMaskInstance = () => {
		Inputmask(DATE_TIME_INPUTMASK_CONFIG).mask(this.nativeInput);
	};

	private onInputHandler = ({ target }: InputEvent) => {
		const input = target as HTMLInputElement | null;
		if (!isNil(input) && input?.value !== this.value) {
			this.textChange.emit(input.value || '');
		}
	};

	private onBlurHandler = ({ target }: FocusEvent) => {
		this.dateTimeBlur.emit((target as HTMLInputElement).value);
		this.focused = false;
	};

	private onFocusHandler = (event: FocusEvent) => {
		this.focused = true;
		this.inputFocus.emit(event);
	};

	private getValue(): string {
		return (this.value || '').toString();
	}

	render() {
		const id = this.el.getAttribute('id');
		const value = this.getValue();

		return (
			<Host>
				<div class="date-time-input-container">
					<kv-form-label label={this.label} required={this.required} />
					<div
						class={{
							'input-container': true,
							[`input-container--size-${this.size}`]: true
						}}
					>
						<Fragment>
							<input
								id={id}
								ref={input => (this.nativeInput = input as HTMLInputElement)}
								type={EInputFieldType.Text}
								name={this.inputName}
								disabled={this.disabled}
								placeholder={this.placeholder}
								value={value}
								onInput={this.onInputHandler}
								onBlur={this.onBlurHandler}
								onFocus={this.onFocusHandler}
								class={{
									'forced-focus': (this.focused || this.highlighted) && !this.disabled,
									'invalid': this.state === EValidationState.Invalid
								}}
							/>
						</Fragment>
					</div>
					<kv-form-help-text helpText={this.helpText} state={this.state} />
				</div>
			</Host>
		);
	}
}
