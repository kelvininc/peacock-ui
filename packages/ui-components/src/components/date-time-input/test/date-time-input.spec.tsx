import { SpecPage } from '@stencil/core/internal';
import { KvDateTimeInput } from '../date-time-input';
import { newSpecPage } from '@stencil/core/testing';

describe('Date Time Input (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvDateTimeInput],
				html: `<kv-date-time-input></kv-date-time-input>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has a label', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvDateTimeInput],
				html: `<kv-date-time-input label="Date Time Label"></date-time-input>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
