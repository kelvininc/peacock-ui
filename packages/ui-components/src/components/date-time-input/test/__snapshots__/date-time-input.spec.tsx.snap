// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Time Input (unit tests) when has a label should match the snapshot 1`] = `
<kv-date-time-input label="Date Time Label" placeholder="" state="none" value="">
  <div class="date-time-input-container">
    <kv-form-label label="Date Time Label"></kv-form-label>
    <div class="input-container input-container--size-large">
      <input placeholder="" type="text" value="">
    </div>
    <kv-form-help-text state="none"></kv-form-help-text>
  </div>
</kv-date-time-input>
`;

exports[`Date Time Input (unit tests) when uses default props should match the snapshot 1`] = `
<kv-date-time-input placeholder="" state="none" value="">
  <div class="date-time-input-container">
    <kv-form-label></kv-form-label>
    <div class="input-container input-container--size-large">
      <input placeholder="" type="text" value="">
    </div>
    <kv-form-help-text state="none"></kv-form-help-text>
  </div>
</kv-date-time-input>
`;
